### PR TITLE
Add ONNX dynamo metadata documentation

### PR DESCRIPTION
**Pull request by bmehta001**
Describe auto-generated metadata when calling torch.onnx.export

cc @justinchuby

### CODE REVIEW

**Comment by bmehta001** [comment]
@titaiwangms @justinchuby @gramalingam @xadupre

**Comment by justinchuby** [comment]
I think we intended to describe the `metadata_props` fields on the onnx model, graph and individual nodes.

File: docs/source/onnx_dynamo.md [hunk 1]
```diff
@@ -144,6 +144,16 @@ to resolve the issue.
 onnx_dynamo_memory_usage
 ```
 
+## Metadata
+
+The `ExportedProgram` class is a representation of a PyTorch model that has been exported to ONNX format. It contains the model's computation graph, which includes the operations performed on the input tensors, their types, and shapes. The `ExportedProgram` class consists of:
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]: We don't need to describe the ExportedProgram class. Instead, I would describe the fields in the ir.Model metadata_props and in graph and individual nodes. Describe what information is recorded and in what format.

**Comment by justinchuby** [comment]
@bmehta001 we may want this to catch the 2.8 release

File: docs/source/onnx_dynamo.md (hunk 1 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: Ah thanks, I was only seeing the props from when I printed out the model. This covered the graph_signature/range_constraints metadata props for the graph, so now adding the node/input/output/initializer metadata props

File: docs/source/onnx_dynamo.md [hunk 2]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+
+  *Example:*
+  `_empty_nn_module_stack_from_metadata_hook: _empty_nn_module_stack_from_metadata_hook/sym_size_int_720: aten.sym_size.int`
+
+- **pkg.torch.onnx.class_hierarchy**
+
+  A list of class names representing the hierarchy of modules leading to this node.
+  
+  *Example:*
+  `['_empty_nn_module_stack_from_metadata_hook', 'aten.sym_size.int']`
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```
  `['SomeModuleName(?)', 'aten.sym_size.int']`
```

maybe more representative to show that this is the nn module class name?

File: docs/source/onnx_dynamo.md [hunk 3]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+  - "GRADIENT_TO_PARAMETER": Gradient output for a parameter.
+  - "GRADIENT_TO_USER_INPUT": Gradient output for a user input.
+  - "USER_INPUT_MUTATION": Indicates a user input was mutated.
+  - "TOKEN": A token output.
+
+Each initializer, input, output has the following metadata:
+
+- **pkg.torch.onnx.original_node_name**
+  
+  The original name of the node in the PyTorch FX graph that produced this initializer. This helps trace initializers back to their source in the original model.
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```suggestion
  The original name of the node in the PyTorch FX graph that produced this value. This helps trace initializers back to their source in the original model.
```

File: docs/source/onnx_dynamo.md [hunk 4]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+
+  *Example:*
+  `fc1.weight`
+
+## Metadata
+
+The `ExportedProgram` class is a representation of a PyTorch model that has been exported to ONNX format. It contains the model's computation graph, which includes the operations performed on the input tensors, their types, and shapes. The `ExportedProgram` class consists of:
+- The `GraphModule` class, a PyTorch module that represents the exported program. It includes a `forward` method that defines the computation graph.
+  - The `forward` method inputs are defined as parameters with specific tensor types and shapes, such as "f32[8, 8]" for a float32 tensor with shape [8, 8].
+  - The operations within the `forward` method are represented as `torch.ops.aten` calls. Each operation is named (e.g., `torch.ops.aten.conv2d.default`) and includes the inputs and outputs with their respective types and shapes.
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```suggestion
```

File: docs/source/onnx_dynamo.md [hunk 5]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+  - Each operation in the graph is annotated with its source file and line number, providing context for where the operation was defined in the original code (e.g., `# File: ~/python/site-packages/torch/nn/modules/conv.py:554 in forward, code: return self._conv_forward(input, self.weight, self.bias)`).
+- The `Graph signature`, which defines the input and output specifications of the model. The inputs are defined as `InputSpec` objects, which include the kind of input (e.g., `InputKind.PARAMETER` for parameters, `InputKind.USER_INPUT` for user-defined inputs), the argument name, the target (which can be a specific layer in the model), and whether the input is persistent. The outputs are defined as `OutputSpec` objects, which specify the kind of output (e.g., `OutputKind.USER_OUTPUT`) and the argument name.
+- A `Range constraints` dictionary, which specifies any constraints on the input sizes or shapes, such as `s0: VR[2, int_oo]`, indicating that the size of the input tensor can vary between 2 and infinity.
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```suggestion
```

File: docs/source/onnx_dynamo.md [hunk 6]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+Each initializer, input, output has the following metadata:
+
+- **pkg.torch.onnx.original_node_name**
+  
+  The original name of the node in the PyTorch FX graph that produced this initializer. This helps trace initializers back to their source in the original model.
+
+  *Example:*
+  `fc1.weight`
+
+## Metadata
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```suggestion
```

File: docs/source/onnx_dynamo.md [hunk 7]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+  *Example values:*
+  - "USER_OUTPUT": A user-visible output.
+  - "LOSS_OUTPUT": A loss value output.
+  - "BUFFER_MUTATION": Indicates a buffer was mutated.
+  - "GRADIENT_TO_PARAMETER": Gradient output for a parameter.
+  - "GRADIENT_TO_USER_INPUT": Gradient output for a user input.
+  - "USER_INPUT_MUTATION": Indicates a user input was mutated.
+  - "TOKEN": A token output.
+
+Each initializer, input, output has the following metadata:
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```suggestion
Each initialized value, input, output has the following metadata:
```

File: docs/source/onnx_dynamo.md [hunk 8]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+
+  *Example:*
+  `File "torch/fx/passes/runtime_assert.py", line 24, in insert_deferred_runtime_asserts`
+
+- **"pkg.torch.onnx.input_names**
+
+  The inputs for a node:
+
+  *Example:*
+  `[2]`
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```suggestion
  `[2]`
```

can you share where it shows up? Strange to have a number as input

File: docs/source/onnx_dynamo.md [hunk 9]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+  
+  *Example:*
+  `%sym_size_int_720 : [num_users=1] = call_function[target=torch.ops.aten.sym_size.int](args = (%img, 2), kwargs = {})`
+
+- **pkg.torch.onnx.name_scopes**
+
+  A list of name scopes (methods) representing the path to this node in the PyTorch model.
+
+  *Example:*
+  `['_empty_nn_module_stack_from_metadata_hook', 'sym_size_int_720']`
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]: Can we have a more representative example? Potentially just create a model with a single `add` as an example

File: docs/source/onnx_dynamo.md [hunk 10]
```diff
@@ -143,6 +143,124 @@ to resolve the issue.
...
+
+- **pkg.torch.onnx.stack_trace**
```
Comment (gramalingam) [inline docs/source/onnx_dynamo.md:180]: As a user, I feel overwhelmed by the presence of many, similar looking, metadata components. Anyway: if I am trying to locate the context for the source that produced a node, is this "stack_trace" the most useful of all the pieces above? 

I assume it is a stack trace, though the example below shows only one entry from the stack? Is it possible to have an example to show multiple entries of the call stack?

Comment (justinchuby) [reply docs/source/onnx_dynamo.md:180]: > possible to have an example to show multiple entries of the call stack

yeah that would be good

File: docs/source/onnx_dynamo.md (hunk 4 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: Sorry, I think a merge conflict left a lot of the old documentation in here

File: docs/source/onnx_dynamo.md [hunk 11]
```diff
@@ -143,6 +143,109 @@ to resolve the issue.
...
+
+  *Example:*
+  `['', 'add']`
+
+- **pkg.torch.onnx.stack_trace**
+
+  The stack trace from the original code where this node was created, if available.
+
+  *Example:*
+  `File "torch/fx/passes/runtime_assert.py", line 24, in insert_deferred_runtime_asserts`
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]: Can you update this with the simple add model as well? It should be a full stack

```
...
```

File: docs/source/onnx_dynamo.md [hunk 12]
```diff
@@ -143,6 +143,109 @@ to resolve the issue.
...
+
+These metadata fields are stored in the metadata_props attribute of each ONNX node and can be inspected using Netron or programmatically.
+
+The overall ONNX IR graph has the following metadata props:
+
+- **pkg.torch.export.ExportedProgram.graph_signature**
+
+  This property contains a string representation of the graph_signature from the original PyTorch ExportedProgram. The graph signature describes the structure of the model's inputs and outputs and how they map to the ONNX graph. The inputs are defined as `InputSpec` objects, which include the kind of input (e.g., `InputKind.PARAMETER` for parameters, `InputKind.USER_INPUT` for user-defined inputs), the argument name, the target (which can be a specific node in the model), and whether the input is persistent. The outputs are defined as `OutputSpec` objects, which specify the kind of output (e.g., `OutputKind.USER_OUTPUT`) and the argument name.
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md:196]: Can you create a link to the torch.export documentation section?

File: docs/source/onnx_dynamo.md [hunk 13]
```diff
@@ -143,6 +143,109 @@ to resolve the issue.
...
+
+- **pkg.torch.export.ExportedProgram.range_constraints**
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md:200]: Same here: link to torch.export doc page

File: docs/source/onnx_dynamo.md (hunk 8 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: I could not generate a model with this metadata attribute, so I was assuming that the input names used were the same as what was seen in pkg.torch.onnx.fx_node which had args like `args = ([%tensor_x, %input_dict_tensor_x, %input_list_0], 1)`. But looking through the code more closely, I don't think this attribute is ever generated, as there is no path that calls the function that generates it, so I am removing this attribute entirely.

File: docs/source/onnx_dynamo.md [hunk 14]
```diff
@@ -143,6 +143,109 @@ to resolve the issue.
...
+  - "GRADIENT_TO_PARAMETER": Gradient output for a parameter.
+  - "GRADIENT_TO_USER_INPUT": Gradient output for a user input.
+  - "USER_INPUT_MUTATION": Indicates a user input was mutated.
+  - "TOKEN": A token output.
+
+Each initialized value, input, output has the following metadata:
+
+- **pkg.torch.onnx.original_node_name**
+  
+  The original name of the node in the PyTorch FX graph that produced this value. This helps trace initializers back to their source in the original model.
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]: In the case where value was renamed.

File: docs/source/onnx_dynamo.md (hunk 8 above)
Comment (justinchuby) [reply docs/source/onnx_dynamo.md]: Should be here: https://github.com/pytorch/pytorch/blob/f402eed4d9515dee5bb9b11a2b157de57a5988c1/torch/onnx/_internal/exporter/_core.py#L419

Comment (justinchuby) [reply docs/source/onnx_dynamo.md]: You are right this is a path we don't expose to users. Feel free to remove

File: docs/source/onnx_dynamo.md (hunk 2 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: Fixed

File: docs/source/onnx_dynamo.md (hunk 11 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: ` 'pkg.torch.onnx.stack_trace': 'File "/home/bhamehta/onnxscript/DONOTPUSH/simpleadd.py", line 7, in forward\n    return torch.add(x, y)` - I only get 1 line with my SimpleAdd model and a few other simple models I tried, so did not update it yet

Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: If I use a more complicated model, I can get a longer stack trace
```
File "/opt/conda/envs/ptca/lib/python3.10/site-packages/transformers/models/electra/modeling_electra.py", line 1615, in forward
    outputs = self.electra(
  File "/opt/conda/envs/ptca/lib/python3.10/site-packages/transformers/models/electra/modeling_electra.py", line 908, in forward
    hidden_states = self.embeddings_project(hidden_states)
  File "/opt/conda/envs/ptca/lib/python3.10/site-packages/torch/nn/modules/linear.py", line 125, in forward
    return F.linear(input, self.weight, self.bias)
    ```

File: docs/source/onnx_dynamo.md (hunk 14 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: Adding

File: docs/source/onnx_dynamo.md (hunk 9 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: Modified

File: docs/source/onnx_dynamo.md (hunk 10 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md:180]: Yes, will have a longer stack trace

File: docs/source/onnx_dynamo.md (hunk 11 above)
Comment (justinchuby) [reply docs/source/onnx_dynamo.md]: Maybe it's ok then (to keep it short), thanks!

File: docs/source/onnx_dynamo.md (hunk 13 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md:200]: Added

File: docs/source/onnx_dynamo.md (hunk 12 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md:196]: Added

File: docs/source/onnx_dynamo.md (hunk 2 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: Using simple add model

File: docs/source/onnx_dynamo.md (hunk 11 above)
Comment (bmehta001) [reply docs/source/onnx_dynamo.md]: Updated

File: docs/source/onnx_dynamo.md [hunk 15]
```diff
@@ -135,14 +135,121 @@ You can view the exported model using [Netron](https://netron.app/).
...
+  The stack trace from the original code where this node was created, if available.
+
+  *Example:*
+  ```
+  File "simpleadd.py", line 7, in forward\n    return torch.add(x, y)
+  ```
+
+These metadata fields are stored in the metadata_props attribute of each ONNX node and can be inspected using Netron or programmatically.
+
+The overall ONNX IR graph has the following metadata props:
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```suggestion
The overall ONNX graph has the following `metadata_props`:
```

File: docs/source/onnx_dynamo.md [hunk 16]
```diff
@@ -135,14 +135,121 @@ You can view the exported model using [Netron](https://netron.app/).
...
+  *Example:*
+  `['', 'add']`
+
+- **pkg.torch.onnx.stack_trace**
+
+  The stack trace from the original code where this node was created, if available.
+
+  *Example:*
+  ```
+  File "simpleadd.py", line 7, in forward\n    return torch.add(x, y)
```
Comment (justinchuby) [inline docs/source/onnx_dynamo.md]:
```suggestion
  File "simpleadd.py", line 7, in forward
      return torch.add(x, y)
```

**Review by justinchuby — APPROVED** [review]
Thanks!

**Review by titaiwangms — APPROVED** [review]
